### PR TITLE
Feat(authentication): username password provider

### PR DESCRIPTION
**Pull request by pKorsholm**

### CODE REVIEW

File: packages/authentication/package.json [hunk 1]
```diff
@@ -31,7 +31,8 @@
     "prepublishOnly": "cross-env NODE_ENV=production tsc --build && tsc-alias -p tsconfig.json",
     "build": "rimraf dist && tsc --build && tsc-alias -p tsconfig.json",
     "test": "jest --runInBand --bail --forceExit -- src/**/__tests__/**/*.ts",
-    "test:integration": "jest --runInBand --forceExit -- integration-tests/**/__tests__/**/*.ts",
+    "test:integration": "jest --runInBand --forceExit -- ",
+    "test:integration:single": "jest --runInBand --forceExit ",
```
Comment (carlos-r-l-rodrigues) [inline packages/authentication/package.json]: ?

File: packages/authentication/integration-tests/__tests__/services/module/providers.spec.ts [hunk 2]
```diff
@@ -22,6 +23,10 @@ describe("AuthenticationModuleService - AuthProvider", () => {
         schema: process.env.MEDUSA_PRICING_DB_SCHEMA,
       },
     })
+
+    if(service.__hooks?.onApplicationStart) {
+      await service.__hooks.onApplicationStart()
+    }
```
Comment (adrien2p) [inline packages/authentication/integration-tests/__tests__/services/module/providers.spec.ts:29]: todo: I would suggest to start using the medusa app to start modules. you can see how we achieved that in the search pr https://github.com/medusajs/medusa/blob/7de35d84c889070f34dd18255000bb39fba06a72/packages/medusa-test-utils/src/init-modules.ts

Comment (adrien2p) [reply packages/authentication/integration-tests/__tests__/services/module/providers.spec.ts:29]: cc @olivermrbl

File: packages/authentication/package.json [hunk 3]
```diff
@@ -31,7 +31,8 @@
     "prepublishOnly": "cross-env NODE_ENV=production tsc --build && tsc-alias -p tsconfig.json",
     "build": "rimraf dist && tsc --build && tsc-alias -p tsconfig.json",
     "test": "jest --runInBand --bail --forceExit -- src/**/__tests__/**/*.ts",
-    "test:integration": "jest --runInBand --forceExit -- integration-tests/**/__tests__/**/*.ts",
+    "test:integration": "jest --runInBand --forceExit -- ",
```
Comment (adrien2p) [inline packages/authentication/package.json]: question: is this working?

File: packages/authentication/package.json (hunk 1 above)
Comment (adrien2p) [inline packages/authentication/package.json]: todo: should be rm no?

File: packages/authentication/src/providers/username-password.ts [hunk 4]
```diff
@@ -13,8 +18,48 @@ class UsernamePasswordProvider extends AbstractAuthenticationModuleProvider {
...
   }
 
-  async authenticate(userData: Record<string, unknown>) {
-    return {}
+  async authenticate(
+    userData: Record<string, any>
+  ): Promise<AuthenticationResponse> {
+    const { email, password } = userData.body
+
+    if (typeof password !== "string") {
```
Comment (adrien2p) [inline packages/authentication/src/providers/username-password.ts]: todo: use isString

File: packages/authentication/src/providers/username-password.ts [hunk 5]
```diff
@@ -13,8 +18,48 @@ class UsernamePasswordProvider extends AbstractAuthenticationModuleProvider {
     this.authUserSerivce_ = AuthUserService
   }
 
-  async authenticate(userData: Record<string, unknown>) {
-    return {}
+  async authenticate(
+    userData: Record<string, any>
```
Comment (adrien2p) [inline packages/authentication/src/providers/username-password.ts:23]: todo: maybe start typing this

File: packages/authentication/src/providers/username-password.ts [hunk 6]
```diff
@@ -13,8 +18,48 @@ class UsernamePasswordProvider extends AbstractAuthenticationModuleProvider {
...
+    const { email, password } = userData.body
+
+    if (typeof password !== "string") {
+      return {
+        success: false,
+        error: "Password should be a string",
+      }
+    }
+
+    if (typeof email !== "string") {
```
Comment (adrien2p) [inline packages/authentication/src/providers/username-password.ts]: todo: use isString

File: packages/authentication/src/providers/username-password.ts [hunk 7]
```diff
@@ -13,8 +18,48 @@ class UsernamePasswordProvider extends AbstractAuthenticationModuleProvider {
...
+    }
+
+    const authUser = await this.authUserSerivce_.retrieveByProviderAndEntityId(
+      email,
+      UsernamePasswordProvider.PROVIDER
+    )
+
+    const password_hash = authUser.provider_metadata?.password
+
+    if (password_hash && typeof password_hash === "string") {
```
Comment (adrien2p) [inline packages/authentication/src/providers/username-password.ts]: todo: same

File: packages/authentication/src/services/authentication-module.ts [hunk 8]
```diff
@@ -37,6 +43,38 @@ export default class AuthenticationModuleService<
...
+          continue
+        }
+
+        providersToCreate.push({
+          provider: provider.provider,
+          name: provider.displayName,
+        })
+      }
+
+      await this.authProviderService_.create(providersToCreate)
```
Comment (adrien2p) [inline packages/authentication/src/services/authentication-module.ts]: suggestion: Could you move that to a private method and call it here just to not surcharge the hook readability

File: packages/authentication/src/providers/username-password.ts [hunk 9]
```diff
@@ -13,8 +18,48 @@ class UsernamePasswordProvider extends AbstractAuthenticationModuleProvider {
...
+      const buf = Buffer.from(password_hash, "base64")
+
+      const success = await Scrypt.verify(buf, password)
```
Comment (carlos-r-l-rodrigues) [inline packages/authentication/src/providers/username-password.ts:51]: suggestion: experiment running that with `worker_threads` (it doesn't have to be in this PR)

File: packages/authentication/src/services/authentication-module.ts [hunk 10]
```diff
@@ -336,7 +375,38 @@ export default class AuthenticationModuleService<
...
   }
 
-  __joinerConfig(): ModuleJoinerConfig {
-    return joinerConfig
+  protected getRegisteredAuthenticationProvider(
+    provider: string
+  ): AbstractAuthenticationModuleProvider {
+    let containerProvider: AbstractAuthenticationModuleProvider
+    try {
+      containerProvider = this.__container__[`auth_provider_${provider}`]
```
Comment (carlos-r-l-rodrigues) [inline packages/authentication/src/services/authentication-module.ts:360]: it's the same behavior, but maybe we should explicitly call `.resolve` where using `this.__container__` ?

**Review by carlos-r-l-rodrigues — APPROVED** [review]
LGTM

File: packages/authentication/src/services/authentication-module.ts (hunk 10 above)
Comment (adrien2p) [reply packages/authentication/src/services/authentication-module.ts:360]: unfortunately, there is no resolve on the proxy. so this is the right way

File: packages/authentication/integration-tests/__tests__/services/module/providers.spec.ts (hunk 2 above)
Comment (olivermrbl) [reply packages/authentication/integration-tests/__tests__/services/module/providers.spec.ts:29]: I am not super happy about these service hooks. The consumer of a module service should never have to worry about the providers being created on load - even when using initialize. 

However, this concern can be discussed and tackled separately from this PR.

File: packages/authentication/src/loaders/providers.ts [hunk 11]
```diff
@@ -17,33 +16,22 @@ export default async ({
...
+  container.register({
+    [`auth_providers`]: asArray(providersToLoad),
+  })
+}
 
-    providersToCreate.push({
-      provider: provider.PROVIDER,
-      name: provider.DISPLAY_NAME,
-    })
+function asArray(
```
Comment (olivermrbl) [inline packages/authentication/src/loaders/providers.ts:30]: question: Seems off to place a generic util like this here

File: packages/authentication/src/migrations/Migration20240115092929.ts [hunk 12]
```diff
@@ -0,0 +1,15 @@
+import { Migration } from '@mikro-orm/migrations';
```
Comment (olivermrbl) [inline packages/authentication/src/migrations/Migration20240115092929.ts:1]: comment: You should be able to omit the migration entirely until we have finalized the module. The tests will read the models and do not depend on migration files.

File: packages/authentication/src/models/auth-user.ts [hunk 13]
```diff
@@ -1,25 +1,32 @@
...
+  Unique,
 } from "@mikro-orm/core"
+
 import AuthProvider from "./auth-provider"
+import { generateEntityId } from "@medusajs/utils"
 
 type OptionalFields = "provider_metadata" | "app_metadata" | "user_metadata"
 
 @Entity()
+@Unique({ properties: ["provider","entity_id" ], name: "IDX_auth_user_provider_entity_id" })
```
Comment (olivermrbl) [inline packages/authentication/src/models/auth-user.ts:20]: question: Does this need linting?

**Review by olivermrbl — APPROVED** [review]
LGTM

File: packages/authentication/integration-tests/__tests__/services/module/providers.spec.ts (hunk 2 above)
Comment (carlos-r-l-rodrigues) [reply packages/authentication/integration-tests/__tests__/services/module/providers.spec.ts:29]: This is needed for modules that require something to be executed after all modules are loaded and the application is running.
It was originally created at the index/search module where it reads all the exported joiner configs of loaded modules.

This is not executed by the consumer of the modules. MedusaApp calls it after it has completed loading everything.

Comment (carlos-r-l-rodrigues) [reply packages/authentication/integration-tests/__tests__/services/module/providers.spec.ts:29]: As we discussed a few times, at some point we need to normalize all entrypoints to be MedusaApp, and discourage the use of `initialize` directly

**Pull request merged**